### PR TITLE
feat: add link unfurling template [WIP]

### PR DESCRIPTION
**Pull request by yukun-dong**
https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_sprints/taskboard/Compute%20Team/Microsoft%20Teams%20Extensibility/CY23-5.1?workitem=19843603

### CODE REVIEW

File: templates/scenarios/ts/link-unfurling/appPackage/manifest.json.tpl [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    "description": {
+        "short": "short description of {{appName}}",
+        "full": "full description of {{appName}}"
+    },
+    "accentColor": "#FFFFFF",
+    "bots": [],
+    "composeExtensions": [
+        {
+            "botId": "${{BOT_ID}}",
+            "commands": [
```
Comment (yukun-dong) [inline templates/scenarios/ts/link-unfurling/appPackage/manifest.json.tpl:30]: The `composeExtensions` must have at least 1 commands. We can discuss this offline.

File: templates/scenarios/ts/link-unfurling/linkUnfurlingBot.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import {
+  TeamsActivityHandler,
+  CardFactory,
+  TurnContext,
+  MessagingExtensionResponse,
+  AppBasedLinkQuery,
+} from "botbuilder";
+import card from "./card.json";
+
+export class LinkUnfurlingBot extends TeamsActivityHandler {
```
Comment (yukun-dong) [inline templates/scenarios/ts/link-unfurling/linkUnfurlingBot.ts:10]: I used class name from https://github.com/OfficeDev/Microsoft-Teams-Samples/blob/main/samples/msgext-link-unfurling/nodejs/bots/teamsLinkUnfurlingBot.js, but I'm thinking about removing bot concept in this template. Should we use name such as "LinkUnfurlingApp"?

Comment (JerryYangKai) [reply templates/scenarios/ts/link-unfurling/linkUnfurlingBot.ts:10]: LinkUnfurlingBot is OK for me.

File: templates/scenarios/ts/link-unfurling/appPackage/manifest.json.tpl (hunk 1 above)
Comment (JerryYangKai) [reply templates/scenarios/ts/link-unfurling/appPackage/manifest.json.tpl:30]: Got it.

**Review by JerryYangKai — APPROVED** [review]

File: templates/scenarios/ts/link-unfurling/package.json.tpl [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        "type": "git",
+        "url": "https://github.com"
+    },
+    "dependencies": {
+        "botbuilder": "^4.18.0",
+        "restify": "^10.0.0"
+    },
+    "devDependencies": {
+        "@types/restify": "^8.5.5",
+        "@types/node": "^14.0.0",
```
Comment (Siglud) [inline templates/scenarios/ts/link-unfurling/package.json.tpl:29]: We already use node 18 by default, why still use @types/node 14 here?

Comment (yukun-dong) [reply templates/scenarios/ts/link-unfurling/package.json.tpl:29]: Good catch. Let me update.

File: templates/scenarios/ts/link-unfurling/linkUnfurlingBot.ts (hunk 2 above)
Comment (MuyangAmigo) [reply templates/scenarios/ts/link-unfurling/linkUnfurlingBot.ts:10]: Let's use `LinkUnfurlingApp`, Bot isn't a terminology on Outlook.

File: templates/scenarios/ts/link-unfurling/card.json [hunk 4]
```diff
@@ -0,0 +1,45 @@
+{
```
Comment (MuyangAmigo) [inline templates/scenarios/ts/link-unfurling/card.json:1]: Are we getting rid of `src` folder? Also can we follow the pattern for notification/command bot where the adaptive cards are put under `/adaptiveCards` path and probably name it as `helloworldLinkUnfurlingCard.json` or something similar.

File: templates/scenarios/ts/link-unfurling/README.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+              type: "setCachePolicy",
+              value: '{"type":"no-cache"}',
+            },
+          ],
+        }
+```
+After removing this, the link unfurling result will be cached in Teams for 30 minutes. 
+
+Please refer to [this document](https://learn.microsoft.com/en-us/microsoftteams/platform/messaging-extensions/how-to/link-unfurling?tabs=desktop%2Cjson%2Cadvantages#remove-link-unfurling-cache) for more details.
+
```
Comment (MuyangAmigo) [inline templates/scenarios/ts/link-unfurling/README.md:78]: Can we reformat the README? Something similar with command bot. We need to include:
* An overview of this template.
* Get Started that includes Prerequisites, and how to run this app in Teams/Outlook.
* What's included in the template, a folder structure walkthrough.
* Extend this template (Customization Scenario)
  * How to use Zero Install
  * How to add stage view / task module / adaptive card actions
  * How to customize adaptive cards.
  * How to handle regular expressions with unfurled links? 
  * etc...

File: templates/scenarios/ts/link-unfurling/appPackage/manifest.json.tpl (hunk 1 above)
Comment (eriolchan) [reply templates/scenarios/ts/link-unfurling/appPackage/manifest.json.tpl:30]: That's a little weird at the first glance. Is there any Teams sample can refer to?

Comment (eriolchan) [reply templates/scenarios/ts/link-unfurling/appPackage/manifest.json.tpl:30]: Although I find this in Teams Sample https://github.com/OfficeDev/Microsoft-Teams-Samples/blob/main/samples/msgext-link-unfurling/nodejs/bots/teamsLinkUnfurlingBot.js. It include both link unfurling and search in link unfurling sample.

Can we check with Junjie to see if we need to merge link unfurling with search query in one scenario, or ask the recommendation from ME team?

File: templates/scenarios/ts/link-unfurling/linkUnfurlingBot.ts (hunk 2 above)
Comment (eriolchan) [reply templates/scenarios/ts/link-unfurling/linkUnfurlingBot.ts]: Agree with Junjie, we'd better to use LinkUnfurlingApp

File: templates/scenarios/ts/link-unfurling/card.json (hunk 4 above)
Comment (eriolchan) [reply templates/scenarios/ts/link-unfurling/card.json]: Maybe helloWorldCard.json is enough?

File: templates/scenarios/ts/link-unfurling/README.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+- [Node.js](https://nodejs.org/), supported versions: 16, 18
+- A Microsoft 365 account. If you do not have Microsoft 365 account, apply one from [Microsoft 365 developer program](https://developer.microsoft.com/en-us/microsoft-365/dev-program)
+- [Teams Toolkit Visual Studio Code Extension](https://aka.ms/teams-toolkit) version 5.0.0 and higher or [TeamsFx CLI](https://aka.ms/teamsfx-cli)
+
+## Debug
+- From Visual Studio Code: Click `Run and Debug` panel.
+- Select a target Microsoft application where the link unfurling app runs: `Debug in Teams`, `Debug in Outlook` and click the `Run and Debug` green arrow button.
+- From TeamsFx CLI: 
+  - Install [ngrok](https://ngrok.com/download) and start your local tunnel service by running the command `ngrok http 3978`.
+  - In the `env/.env.local` file, fill in the values for `BOT_DOMAIN` and `BOT_ENDPOINT` with your ngrok URL.
```
[the remark this answers is not in the document]
Comment (yukun-dong) [reply templates/scenarios/ts/link-unfurling/README.md:26]: CLI still uses ngrok for now.

File: templates/scenarios/ts/link-unfurling/README.md (hunk 5 above)
Comment (yukun-dong) [reply templates/scenarios/ts/link-unfurling/README.md:125]: updated.

File: templates/scenarios/ts/link-unfurling/README.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+| `src/index.ts` | Application entry point and `restify` handlers |
+| `src/linkUnfurlingApp.ts`| The teams activity handler |
+| `src/adaptiveCards/helloWorldCard.json` | The adaptive card |
+
+## Extend this template
+
+This section introduces how to customize or extend this template, including:
+
+- [How to use Zero Install Link Unfurling in Teams](https://github.com/OfficeDev/TeamsFx/wiki/How-to-extend-Link-Unfurling-template#how-to-use-zero-install-link-unfurling-in-teams)
+- [How to add link unfurling cache in Teams](https://github.com/OfficeDev/TeamsFx/wiki/How-to-extend-Link-Unfurling-template#how-to-add-link-unfurling-cache-in-teams)
```
Comment (MuyangAmigo) [inline templates/scenarios/ts/link-unfurling/README.md:45]: Please change github wiki urls to an aka link.

Comment (yukun-dong) [reply templates/scenarios/ts/link-unfurling/README.md]: updated.

**Review by MuyangAmigo — DISMISSED** [review]

**Review by MuyangAmigo — APPROVED** [review]

**Pull request merged**